File: ReactAndroid/src/main/java/com/facebook/react/animated/DiffClampAnimatedNode.java
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * <p>This source code is licensed under the MIT license found in the LICENSE file in the root
 * directory of this source tree.
 */
package com.facebook.react.animated;

import com.facebook.react.bridge.JSApplicationCausedNativeException;
import com.facebook.react.bridge.ReadableMap;

/*package*/ class DiffClampAnimatedNode extends ValueAnimatedNode {
  private final NativeAnimatedNodesManager mNativeAnimatedNodesManager;
  private final int mInputNodeTag;
  private final double mMin;
  private final double mMax;

  private double mLastValue;

  public DiffClampAnimatedNode(
      ReadableMap config, NativeAnimatedNodesManager nativeAnimatedNodesManager) {
    mNativeAnimatedNodesManager = nativeAnimatedNodesManager;
    mInputNodeTag = config.getInt("input");
    mMin = config.getDouble("min");
    mMax = config.getDouble("max");

    mValue = mLastValue = 0;
  }

  @Override
  public void update() {
    double value = getInputNodeValue();

    double diff = value - mLastValue;
    mLastValue = value;
    mValue = Math.min(Math.max(mValue + diff, mMin), mMax);
  }

  private double getInputNodeValue() {
    AnimatedNode animatedNode = mNativeAnimatedNodesManager.getNodeById(mInputNodeTag);
    if (animatedNode == null || !(animatedNode instanceof ValueAnimatedNode)) {
      throw new JSApplicationCausedNativeException(
          "Illegal node ID set as an input for " + "Animated.DiffClamp node");
    }

    return ((ValueAnimatedNode) animatedNode).getValue();
  }
}

File: packages/react-native/ReactCxxPlatform/react/renderer/animated/nodes/DiffClampAnimatedNode.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/*
 * Adapted from react-native-windows under the MIT license.
 */

#include "DiffClampAnimatedNode.h"

#include <react/renderer/animated/NativeAnimatedNodesManager.h>

namespace facebook::react {

DiffClampAnimatedNode::DiffClampAnimatedNode(
    Tag tag,
    const folly::dynamic& config,
    NativeAnimatedNodesManager& manager)
    : ValueAnimatedNode(tag, config, manager),
      inputNodeTag_(static_cast<Tag>(getConfig()["input"].asDouble())),
      min_(getConfig()["min"].asDouble()),
      max_(getConfig()["max"].asDouble()) {}

void DiffClampAnimatedNode::update() {
  if (const auto node =
          manager_->getAnimatedNode<ValueAnimatedNode>(inputNodeTag_)) {
    const auto value = node->getValue();
    const auto diff = value - lastValue_;
    lastValue_ = value;
    setRawValue(std::clamp(this->getValue() + diff, min_, max_));
  }
}

} // namespace facebook::react
